src/storage/src/source/healthcheck.rs

```diff
@@ -0,0 +1,809 @@
+            let row = source_data
+                .expect("failed to deserialize row")
+                .0
+                .expect("status collection should not have errors");
+            let row_vec = row.unpack();
+            let row_source_id = row_vec[2].unwrap_str();
+            let row_worker_id = row_vec[5].unwrap_int64();
+            let row_status = row_vec[7].unwrap_str();
+
+            if self.source_id == row_source_id && self.worker_id == row_worker_id as usize {
+                self.current_status = SourceStatus::try_from(row_status).expect("invalid status");
+            }
+        }
+    }
+
+    fn prepare_row_update(
+        &self,
+        status_update: &SourceStatusUpdate,
+        ts: u64,
+    ) -> Vec<((SourceData, ()), u64, i64)> {
+        let timestamp = NaiveDateTime::from_timestamp(
+            (ts / 1000)
+                .try_into()
+                .expect("timestamp seconds does not fit into i64"),
+            (ts % 1000 * 1_000_000)
+                .try_into()
+                .expect("timestamp millis does not fit into a u32"),
+        )
+        .into();
+        let source_name = Datum::String(&self.source_name);
+        let source_id = Datum::String(&self.source_id);
+        let source_type: Datum = self.source_type.into();
+        let upstream_name: Datum = self.upstream_name.as_deref().into();
+        let worker_id =
+            Datum::Int64(i64::try_from(self.worker_id).expect("worker_id does not fit into i64"));
+        let worker_count = Datum::Int64(
+            i64::try_from(self.worker_count).expect("worker_count does not fit into i64"),
+        );
+        let status = Datum::String(status_update.status.name());
+        let error = status_update.error.as_deref().into();
+        let metadata = Datum::Null;
+        let row = Row::pack_slice(&[
+            timestamp,
+            source_name,
+            source_id,
+            source_type,
+            upstream_name,
+            worker_id,
+            worker_count,
+            status,
+            error,
+            metadata,
+        ]);
+
+        vec![((SourceData(Ok(row)), ()), ts, 1)]
+    }
+}
+
+/// Identify the state a worker for a given source can be at a point in time
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum SourceStatus {
+    /// Initial state of a Source during initialization.
+    Setup,
+    /// Intended to be the state while the `storaged` process is initializing itself
+    /// Pushed by the Healthchecker on creation.
+    Starting,
+    /// State indicating the source is running fine. Pushed automatically as long
+    /// as rows are being consumed.
+    Running,
+    /// Represents a stall in the ingestion process that might get resolved.
+    /// Existing data is still available and queryable.
+    Stalled,
+    /// Represents a irrecoverable failure in the pipeline. Data from this collection
+    /// is not queryable any longer. The only valid transition from Failed is Dropped.
+    Failed,
```

Comment:
Do you have ideas about when a source would ever be in this state? I naively thought users could always keep querying what's already there in persist.

Reply:
They can't if we ever push something to error collection of a source! That said, the most common case is for decoding errors, which are unfortunately not captured here anyway, as this happens further down in the processing pipeline, but I was thinking we could potentially do something there later on.

I think [this Postgres source codepath](https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/source/postgres.rs#L325-L328) is the main relevant way this can happen right now, but I haven't reproduced it locally yet.